### ISSUE TITLE
prterun found 0 available slots

### ISSUE THREAD

**Issue by ggouaillardet** [openpmix/prrte#965]
This is related to open-mpi/ompi#8962.

I could reproduce the issue with the latest `prrte` (`master` branch) and `openpmix` from the `v4.1` branch.

My hostname is `hpcics01` and has an other interface whose host name is `hpcics01-nfs`

```
$ `which prterun` --host hpcics01 hostname
hpcics01.kobe.rist.or.jp
```

works as expected.

```
$ `which prterun` --host hpcics01-nfs hostname
hpcics01.kobe.rist.or.jp
```

seems to work to, but if we look under the cover, `prterun` issues a `ssh hpcics01-nfs ... prted ...` which is suspicious
(e.g. it seems `prrte` could not figure out `hpcics01` and `hpcics01-nfs` are the same host).

now with the FQDN

```
$ `which prterun` --host hpcics01-nfs.kobe.rist.or.jp hostname
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 1
slots that were requested by the application:

  hostname

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the PRTE term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which PRTE processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, PRTE defaults to the number of processor cores

In all the above cases, if you want PRTE to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --map-by :OVERSUBSCRIBE option to ignore the
number of available slots when deciding the number of processes to
launch.
----------------------------------------------------------------------
```

the command fails with a bizarre error message (!)

FWIW, `prterun` did under the hood `ssh hpcics01-nfs ... prted ...`
(note the FQDN was dropped in the SSH command)

**Comment by rhc54** [comment]
@ggouaillardet Please see if https://github.com/openpmix/prrte/pull/966 resolves this problem, and the corresponding issue in OMPI.

**Comment by ggouaillardet** [comment]
Thanks @rhc54 

It fixes the weird behavior (e.g. no more `ssh ...prted ...` on the same host when an alias is used), but I still get the same error

> There are not enough slots available in the system to satisfy the 1 slots that were requested by the application:

**Comment by rhc54** [comment]
I take it you are running `$ prterun --host hpcics01-nfs.kobe.rist.or.jp hostname` - yes? Is this in an allocation? In other words, is there an allocation or default hostfile involved here?

**Comment by rhc54** [comment]
Afraid I cannot replicate the problem, but I also don't have a node with named interfaces. Perhaps it would help if you ran the command with `--prtemca ras_base_verbose 10` so we can see the allocator work?

**Comment by ggouaillardet** [comment]
@rhc54 the command is correct (same behavior when using `--host hpcics01-nfs`). I SSH'ed directly into the compute node, and have no default hostfile

Here is the log of the failure
```
[gilles@hpcics03 ~]$ ~/debuginfo/prrte-pmix-v4.1/bin/prterun -n 1 --host hpcics03-nfs --prtemca ras_base_verbose 10 hostname
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_register: registering framework ras components
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_register: found loaded component simulator
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_register: component simulator register function successful
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_register: found loaded component slurm
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_register: component slurm register function successful
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_open: opening ras components
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_open: found loaded component simulator
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_open: found loaded component slurm
[hpcics03.kobe.rist.or.jp:102260] mca: base: components_open: component slurm open function successful
[hpcics03.kobe.rist.or.jp:102260] mca:base:select: Auto-selecting ras components
[hpcics03.kobe.rist.or.jp:102260] mca:base:select:(  ras) Querying component [simulator]
[hpcics03.kobe.rist.or.jp:102260] mca:base:select:(  ras) Querying component [slurm]
[hpcics03.kobe.rist.or.jp:102260] mca:base:select:(  ras) No component selected!
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] ras:base:allocate
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] ras:base:allocate nothing found in module - proceeding to hostfile
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] ras:base:allocate adding dash_hosts
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: parsing args hpcics03-nfs
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: working node hpcics03-nfs
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: added node hpcics03.kobe.rist.or.jp to list - slots 0
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: adding node hpcics03.kobe.rist.or.jp with 1 slots to final list
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] ras:base:node_insert inserting 1 nodes
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] ras:base:node_insert updating HNP [hpcics03.kobe.rist.or.jp] info to 1 slots

======================   ALLOCATED NODES   ======================
	hpcics03.kobe.rist.or.jp: slots=1 max_slots=0 slots_inuse=0 state=UP
	flags: DAEMON_LAUNCHED:LOCATION_VERIFIED:SLOTS_GIVEN
=================================================================
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: parsing args hpcics03-nfs
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: working node hpcics03-nfs
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: added node hpcics03 to list - slots 0
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: adding node hpcics03 with 1 slots to final list

======================   ALLOCATED NODES   ======================
	hpcics03.kobe.rist.or.jp: slots=1 max_slots=0 slots_inuse=0 state=UP
	flags: DAEMON_LAUNCHED:LOCATION_VERIFIED:SLOTS_GIVEN
=================================================================
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] ras:base:allocate
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] ras:base:allocate allocation already read
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: parsing args hpcics03-nfs
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: working node hpcics03-nfs
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: added node hpcics03 to list - slots 0
[hpcics03.kobe.rist.or.jp:102260] [prterun-hpcics03.kobe.rist.or.jp-102260@0,0] dashhost: adding node hpcics03 with 1 slots to final list
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 1
slots that were requested by the application:

  hostname

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the PRTE term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which PRTE processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, PRTE defaults to the number of processor cores

In all the above cases, if you want PRTE to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --map-by :OVERSUBSCRIBE option to ignore the
number of available slots when deciding the number of processes to
launch.
```

**Comment by rhc54** [comment]
Well that all looked correct - let's add `--prtemca rmaps_base_verbose 5` and watch the mapper

**Comment by ggouaillardet** [comment]
yep, here it is:

```
[gilles@hpcics03 ~]$ ~/debuginfo/prrte-pmix-v4.1/bin/prterun -n 1 --host hpcics03-nfs --prtemca rmaps_base_verbose 5 hostname
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: checking available component mindist
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: Querying component [mindist]
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: checking available component ppr
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: Querying component [ppr]
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: checking available component rank_file
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: Querying component [rank_file]
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: checking available component round_robin
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: Querying component [round_robin]
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: checking available component seq
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:select: Querying component [seq]
[hpcics03.kobe.rist.or.jp:102385] [prterun-hpcics03.kobe.rist.or.jp-102385@0,0]: Final mapper priorities
[hpcics03.kobe.rist.or.jp:102385] 	Mapper: ppr Priority: 90
[hpcics03.kobe.rist.or.jp:102385] 	Mapper: seq Priority: 60
[hpcics03.kobe.rist.or.jp:102385] 	Mapper: mindist Priority: 20
[hpcics03.kobe.rist.or.jp:102385] 	Mapper: round_robin Priority: 10
[hpcics03.kobe.rist.or.jp:102385] 	Mapper: rank_file Priority: 0
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps: mapping job prterun-hpcics03.kobe.rist.or.jp-102385@1
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps: setting mapping policies for job prterun-hpcics03.kobe.rist.or.jp-102385@1 nprocs 1 inherit TRUE
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps[452] mapping not given - using bycore
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps[635] binding not given - using bycore
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:ppr: job prterun-hpcics03.kobe.rist.or.jp-102385@1 not using ppr mapper PPR NULL policy PPR NOTSET
[hpcics03.kobe.rist.or.jp:102385] [prterun-hpcics03.kobe.rist.or.jp-102385@0,0] rmaps:seq called on job prterun-hpcics03.kobe.rist.or.jp-102385@1
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:seq: job prterun-hpcics03.kobe.rist.or.jp-102385@1 not using seq mapper
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:mindist: job prterun-hpcics03.kobe.rist.or.jp-102385@1 not using mindist mapper
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:rr: mapping job prterun-hpcics03.kobe.rist.or.jp-102385@1
[hpcics03.kobe.rist.or.jp:102385] [prterun-hpcics03.kobe.rist.or.jp-102385@0,0] using dash_host hpcics03-nfs
[hpcics03.kobe.rist.or.jp:102385] [prterun-hpcics03.kobe.rist.or.jp-102385@0,0] node hpcics03.kobe.rist.or.jp has 0 slots available
[hpcics03.kobe.rist.or.jp:102385] AVAILABLE NODES FOR MAPPING:
[hpcics03.kobe.rist.or.jp:102385]     node: hpcics03.kobe.rist.or.jp daemon: 0 slots_available: 0
[hpcics03.kobe.rist.or.jp:102385] [prterun-hpcics03.kobe.rist.or.jp-102385@0,0] Starting bookmark at node hpcics03.kobe.rist.or.jp
[hpcics03.kobe.rist.or.jp:102385] [prterun-hpcics03.kobe.rist.or.jp-102385@0,0] Starting at node hpcics03.kobe.rist.or.jp
[hpcics03.kobe.rist.or.jp:102385] mca:rmaps:rr: mapping no-span by Core for job prterun-hpcics03.kobe.rist.or.jp-102385@1 slots 0 num_procs 1
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 1
slots that were requested by the application:

  hostname

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the PRTE term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which PRTE processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, PRTE defaults to the number of processor cores

In all the above cases, if you want PRTE to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --map-by :OVERSUBSCRIBE option to ignore the
number of available slots when deciding the number of processes to
launch.
--------------------------------------------------------------------------
```

**Comment by ggouaillardet** [comment]
Here is a root cause in `prte_node_match()`. It returns false because `strlen(hostname) != strlen(alias)` even before checking aliases.

When invoked on `rokko` with `prterun --host hpcifs-nfs hostname`
/* `hpcifs-nfs` is the name of an other interface */

![prte_node_match](https://user-images.githubusercontent.com/7233259/118941749-02986080-b98d-11eb-910e-7789b625b17f.png)

**Comment by ggouaillardet** [comment]
FWIW, if I revamp this test to go a bit further, it still does not work because `hpcifs-nfs` is **not** in the aliases of the node.

**Comment by ggouaillardet** [comment]
Thanks @rhc54, the issue is now fixed by #966 and #969
